radix-transactions/src/manifest/static_resource_movements_visitor.rs

```diff
@@ -0,0 +1,2171 @@
+use super::*;
+use crate::model::*;
+use radix_common::prelude::*;
+use radix_engine_interface::blueprints::account::*;
+use radix_engine_interface::blueprints::locker::*;
+use std::ops::*;
+use traversal::*;
+use typed_invocations::*;
+
+/// A [`ManifestInterpretationVisitor`] that statically tracks the resources in the worktop and
+/// reports the account withdraws and deposits made.
+pub struct StaticResourceMovementsVisitor {
+    // TODO: How can we do this better? How can we abstract this so that it just does the static
+    // asset movement and make the account stuff added on top?
+    account_deposits: IndexMap<ComponentAddress, Vec<AccountDeposit>>,
+    account_withdraws: IndexMap<ComponentAddress, Vec<AccountWithdraw>>,
+    /// The resource content of the worktop.
+    worktop_fungible_contents: IndexMap<FungibleResourceAddress, FungibleBounds>,
```

Comment:
Having the bucket contents not be separated into two state fields for fungible and non-fungible made the implementation much simpler. Also, I wanted to avoid the case where I needed to lookup the bucket contents in two of the state fields when presented with a `ManifestBucket` because we don't know if it's fungible or not.

Reply:
Yup, I was suggesting we have `worktop_contents: IndexMap<ResourceAddress, Bounds>,` like buckets :)

Reply:
It is primarily because of the logic that we use to add resources to the worktop.

Say that we store the worktop content as the following:

```rust
type WorktopContent = IndexMap<ResourceAddress, ResourceBounds>;

pub enum ResourceBounds {
    Fungible(ResourceAddress, FungibleBounds),
    NonFungible(ResourceAddress, NonFungibleBounds),
}
```

If we have invocations return `ResourceBounds` then the logic to add the invocation output to the worktop content will look something like the following:

```rust
match (worktop_content, other) {
    (Self::Fungible { .. }, Self::Fungible { .. }) => todo!(),
    (Self::NonFungible { .. }, Self::NonFungible { .. }) => todo!(),
    // PROBLEM:
    (Self::Fungible { .. }, Self::NonFungible { .. }) => todo!(),
    (Self::NonFungible { .. }, Self::Fungible { .. }) => todo!(),
}
```

My main issue is the final two arms of the match statement which we have to make `unreachable!()` if we use this pattern since we can't combine fungible and non-fungible resource bounds together as a resource can't be both at the same time.

This separation of fungible content and non-fungible content avoids the above.